dtm/clk_src.h

```diff
@@ -0,0 +1,162 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+#pragma once
+#ifndef __MOTR_DTM_CLK_SRC_H__
+#define __MOTR_DTM_CLK_SRC_H__
+
+#include "xcode/xcode.h" /* xcode attrs */
+#include "lib/time.h" /* m0_time_t */
+
+/* DTM clock source overview
+ * -------------------------
+ *
+ *   Clock source module is designed to be used by DTM0. At this point,
+ * only one type of clock is supported -- physical clock. Such a clock
+ * provides a way to enforce total ordering in a system with synchronised
+ * clocks. However, the interface supports an epoch-based clock source as well.
+ *   See "A.clock.sync" in the DTM0 HLD for additional details.
+ *
+ * Interface
+ * ---------
+ *
+ * CS.TS          -- A data type that represents a point in time (timestamp).
+ * CS.TS.CMP(L,R) -- a 3-way comparison for timestamps; it returns
+ *                   -1 when "left" happened before "right", +1 when
+ *                   "left" happened after "right", and 0 otherwise.
+ * CS.INIT        -- initialise a clock source of the given type.
+ * CS.FINI        -- finalise a lock source.
+ * CS.NOW         -- get the current value of a clock source.
+ * CS.OBSERVED(O) -- make the clock source aware of a timestamp (Other)
+ *                   generated by another instance of the same clock source
+ *                   type.
+ *
+ * Error handling
+ * --------------
+ *
+ *   It is safe to assume that CS.NOW returns "unrecoverable" errors (the caller
+ * should panic if a non-zero value is returned).
+ *   When causality is tracked only by a physical clock (CS_PHYS),
+ * CS.OBSERVED returns -EINVAL if the incomming message "happened" in future.
+ * The caller should stop communication with the clock source that has
+ * generated such a timestamp (notify HA or, at least, panic right here).
+ *
+ * Concurrency
+ * -----------
+ *
+ *   It is safe to call CS.NOW and CS.OBSERVED from any context as long
+ * as the caller owns the corresponding source clock.
+ *   The user must ensure to use only one type of clock source in the system.
+ * The module provides no protection agains such cases.
+ */
+
+
+/**
+ * @defgroup dtm
+ *
+ * @{
+ */
+
+
+enum m0_dtm_cs_types {
+	/* Physical clock.
+	 * Physical clock supports total ordering and requires
+	 * no synchronisation (from DTM0 perspective). But it has
+	 * two drawbacks: simultaneous events are not allowed,
```

Comment:
could you define 'simultaneous events' notion explicitly?

Reply:
I rephrased this comment. Please take a look and let me know if the new definition looks fine.